### PR TITLE
Support Step Load Pattern

### PR DESCRIPTION
**Pull request by delulu**
This PR adds a new feature related to the issue: #1001 

The main idea is to allow you to trigger a load test in Step Load pattern, and you can specify a Step Clients and a Step Duration to monitor your server's performance with different user load.

For example if you specify the swarming plan with 
      `Total Clients of 50, Hatch Rate of 1, Step Clients of 5 and Step Duration of 5m`
click the button and it will trigger a load test with increasing the user load by 5 every 5 mins.

At last you will get the metrics charts as below after 1h' running:

image

image

### CODE REVIEW

**Comment by myzhan** [comment]
I think hatch rate is already a step with 1s as interval.

**Comment by reedstrm** [comment]
How much of the desired feature can be had by allowing control of the length of the hatch-step? Instead of #/sec label it "# spawned per step" and allow setting the "hatch-step length". In the example case, hatch-rate 5, hatch interval 5m should yield most of the same behavior.

**Comment by reedstrm** [comment]
Seems this code does not work with a single local locust instance:
```[2019-06-17 16:15:59,368] sensei/ERROR/stderr: AttributeError: 'LocalLocustRunner' object has no attribute 'start_stepload'```

**Comment by delulu** [comment]
> 
> 
> Seems this code does not work with a single local locust instance:
> `[2019-06-17 16:15:59,368] sensei/ERROR/stderr: AttributeError: 'LocalLocustRunner' object has no attribute 'start_stepload'`

@reedstrm thank you for trying it out, yes, currently start_stepload is only supported in distributed mode.

**Comment by delulu** [comment]
> 
> 
> How much of the desired feature can be had by allowing control of the length of the hatch-step? Instead of #/sec label it "# spawned per step" and allow setting the "hatch-step length". In the example case, hatch-rate 5, hatch interval 5m should yield most of the same behavior.

I want to make sure this change is backward compatible, so I didn't modify the old interface and added two params for step load support.

**Comment by delulu** [comment]
> 
> 
> I think hatch rate is already a step with 1s as interval.

Actually hatch rate is kinda of a booting rate, and the users/second line chart is kinda of linear. while step load is to increase the user load by step and the users/second line chart is kinda of staircase curve. 

For more reference please go to Proposal #1001.

**Comment by max-rocket-internet** [comment]
I think what's proposed in this PR with the extra options is a little hard to understand?

For example, the graph shown doesn't look any different to something you can produce with the current implementation? Maybe at least update the docs to make it clear

**Comment by delulu** [comment]
> 
> 
> I think what's proposed in this PR with the extra options is a little hard to understand?
> 
> For example, the graph shown doesn't look any different to something you can produce with the current implementation? Maybe at least update the docs to make it clear

actually it's a new way to run load test by step, for more details please go to proposal https://github.com/locustio/locust/issues/1001

**Comment by cyberw** [comment]
Hi @delulu ! 

I dont see the need for three command line flags (you could just remove --step-load). 

Could you make another test case that validates the step pattern (number of clients at a certain point in time for example)?

@heyman What do you think of this change? I think it is a nice feature, but I'm unsure about the implementation, and I think we might need to rethink the web gui as we add more parameters.

File: locust/main.py [hunk 1]
```diff
@@ -469,6 +498,9 @@ def main():
         if not options.no_web:
             logger.error("The --run-time argument can only be used together with --no-web")
             sys.exit(1)
+        if options.slave:
```
Comment (cyberw) [inline locust/main.py]: I think this change is not really related to the issue you are solving? Currently I rely on slaves accepting (and ignoring) this parameter, so I'd prefer you not to change this :)

**Comment by cyberw** [comment]
does LocustRunner.stop() really still work after this change? It specifically looks at hatching_greenlet on line 196. I think the correct way of doing this is using only the hatching_greenlet variable, and not adding a new one.

**Comment by heyman** [comment]
I added a comment in the related #1001 issue: https://github.com/locustio/locust/issues/1001#issuecomment-546004337

(I would really like Github to add a feature for merging issues/PRs)

File: locust/main.py [hunk 2]
```diff
@@ -480,43 +512,51 @@ def timelimit_stop():
...
-                time.sleep(1)
-
-            runners.locust_runner.start_hatching(options.num_clients, options.hatch_rate)
-            main_greenlet = runners.locust_runner.greenlet
-            if options.run_time:
-                spawn_run_time_limit_greenlet()
     elif options.slave:
-        if options.run_time:
-            logger.error("--run-time should be specified on the master node, and not on slave nodes")
-            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (cyberw) [reply locust/main.py]: Aha, ok!

**Comment by delulu** [comment]
@cyberw I've update the test case for step load validation, please have a check.

**Comment by delulu** [comment]
merged with latest master branch and resolved conflicts.

**Comment by cyberw** [comment]
Personally I still prefer always having the UI option present (but hidden, maybe under an "advanced" selection together with host name) instead of explicitly passing a --step-load to enable it in the gui. 

But I'm not a gui person anyway, so I'm ok with merging if you resolve the conflicts.

@heyman Your opinion on this? I agree that we may want a more flexible format in the future, but this does solve an important problem in a quite simple way which will be useful to a lot of ppl.

**Comment by delulu** [comment]
@heyman I've replied to your comment, please take a look at https://github.com/locustio/locust/issues/1001#issuecomment-549237782.

**Comment by delulu** [comment]
@cyberw I'll resolve the conflicts, but for the check failure of codecov/patch, I don't know why this check failed, could you show me a way to fix it?

**Comment by cyberw** [comment]
> @cyberw I'll resolve the conflicts, but for the check failure of codecov/patch, I don't know why this check failed, could you show me a way to fix it?

Ignore it :)

**Comment by cyberw** [comment]
LGTM. I will merge soon if nobody objects (@heyman ?). I would like to improve the UI, but we can do that later.

**Comment by cyberw** [comment]
Thanks for your contribution!

**Comment by cyberw** [comment]
@delulu I just realized this is not documented (outside the help function). Would you consider adding some documentation?

**Comment by delulu** [comment]
> 
> 
> @delulu I just realized this is not documented (outside the help function). Would you consider adding some documentation?

sure, will add it : -)

**Comment by delulu** [comment]
Here is the PR https://github.com/locustio/locust/pull/1235 for the documentation.
